### PR TITLE
feat(fault-injection): support conditional fault injection using nginx variables

### LINKED ISSUE #2511: [discuss]: use fault-injection plugin to implement mock function
### Issue description

We can add new features to the fault-injection plugin for the following reasons:
Currently, the `fault-injection` plugin does not support custom conditions to implement the `mock` function. I think we can add the request header as a conditional judgment in the request (there may be a better way) to realize the mock function.
E.g:

```
"vars": [
     [ "http_mock01", "==", " api01" ],
     [ "http_mock02", "==", " api02" ]
]
```
When there are multiple conditions in `vars`, they have an `or` relationship.
If `vars` passes the verification, fault injection is performed on the request. Otherwise, no processing is done on the request. In order to maintain backward compatibility, the fault injection operation is still performed when the vars condition is not configured.

### PR DESCRIPTION
**Pull request by Firstsawyou**
### What this PR does / why we need it:

We can add new features to the fault-injection plugin for the following reasons:
Currently, the `fault-injection` plugin does not support custom conditions to implement the `mock` function. I think we can add the request header as a conditional judgment in the request (there may be a better way) to realize the mock function.
E.g:

```
"vars": [
     [[ "http_mock01", "==", " api01" ]],
     [[ "http_mock02", "==", " api02" ]]
]
```
If `vars` passes the verification, fault injection is performed on the request. Otherwise, no processing is done on the request. In order to maintain backward compatibility, the fault injection operation is still performed when the vars condition is not configured.

close #2511

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [x] Have you modified the corresponding document?
* [x] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

**Comment by membphis** [comment]
ci failed

**Comment by Firstsawyou** [comment]
> ci failed

I need to adjust the code logic and I will solve the ci problem later.

File: apisix/plugins/fault-injection.lua [hunk 1]
```diff
@@ -15,13 +15,66 @@
...
+local expr = require("resty.expr.v1")
 
 local sleep = core.sleep
 local random = math.random
+local ipairs = ipairs
 
 local plugin_name   = "fault-injection"
 
 
+local vars_schema = {
```
Comment (spacewander) [inline apisix/plugins/fault-injection.lua:27]: I think we need to remove the `vars_schema`.
1. we can do the check better with `expr.new` when checking the schema.
2. when `not`/`or` is supported directly in the expr, this schema is outdated.
3. this schema contains many incorrect hardcoded limitations, like the `maxLength` of operator is `2` (not enough for `has`), and an expression may contain 4 elements if `!` is used, etc.

File: apisix/plugins/fault-injection.lua [hunk 2]
```diff
@@ -64,6 +119,25 @@ local function sample_hit(percentage)
 end
 
 
+local function vars_match(vars, ctx)
+    local match_result
+    for _, var in ipairs(vars) do
+        local expr, err = expr.new(var)
+        if err then
+            core.log.error("vars expression does not match: ", err)
```
Comment (spacewander) [inline apisix/plugins/fault-injection.lua:127]: Should be `failed to create vars expression`, we haven't matched it yet.

File: doc/plugins/fault-injection.md [hunk 3]
```diff
@@ -30,11 +30,43 @@ Fault injection plugin, this plugin can be used with other plugins and will be e
...
+        [ "arg_name","==","jack" ],
+        [ "arg_age","==",18 ]
+    ],
+    [
+        [ "arg_name2","==","allen" ]
+    ]
+]
+```
+
+Indicates that the relationship between the first two expressions is `and`, and the relationship between the first two expressions and the third expression is `or`.
```
Comment (spacewander) [inline doc/plugins/fault-injection.md:68]: What `Indicates`? There is a grammar error in the sentence.

File: doc/plugins/fault-injection.md [hunk 4]
```diff
@@ -30,11 +30,43 @@ Fault injection plugin, this plugin can be used with other plugins and will be e
 | abort.http_status | integer | required    |         | [200, ...] | user-specified http code returned to the client. |
 | abort.body        | string  | optional    |         |            | response data returned to the client. Nginx variable can be used inside, like `client addr: $remote_addr\n`           |
 | abort.percentage  | integer | optional    |         | [0, 100]   | percentage of requests to be aborted.            |
+| abort.vars        | array[] | optional    |         |            | The rules for executing fault injection will only be executed when the rules are matched. `vars` is a list consisting of one or more [[var, operator, val]] elements, like this: [[[var, operator, val],[var, operator, val]], ...]. For example: [[["arg_name", "==", "json"]]], indicating that the current request parameter name is json. The var here is consistent with the naming of Nginx internal variables, so request_uri, host, etc. can also be used; for the operator part, the currently supported operators are ==, ~=, ~~, >, <, in, has and !. For specific usage of operators, please see the `operator-list` part of [lua-resty-expr](https://github.com/api7/lua-resty-expr#operator-list).  |
```
Comment (spacewander) [inline doc/plugins/fault-injection.md:33]: I think the description can be simpler. It has been repeated for too many times.
We can say it is a list of `expression`, which is from the lua-resty-expr. That is all.

File: doc/plugins/fault-injection.md [hunk 5]
```diff
@@ -30,11 +30,43 @@ Fault injection plugin, this plugin can be used with other plugins and will be e
...
 | delay.duration    | number  | required    |         |            | delay time (can be decimal).                     |
 | delay.percentage  | integer | optional    |         | [0, 100]   | percentage of requests to be delayed.            |
+| delay.vars        | array[] | optional    |         |            | Execute the request delay rule, and the request will be delayed only after the rule matches. `vars` is a list consisting of one or more [[var, operator, val]] elements, like this: [[[var, operator, val],[var, operator, val]], ...]. For example: [[["arg_name", "==", "json"]]], indicating that the current request parameter name is json. The var here is consistent with the naming of Nginx internal variables, so request_uri, host, etc. can also be used; for the operator part, the currently supported operators are ==, ~=, ~~, >, <, in, has and !. For specific usage of operators, please see the `operator-list` part of [lua-resty-expr](https://github.com/api7/lua-resty-expr#operator-list).  |
 
 Note: One of `abort` and `delay` must be specified.
 
+`vars` is composed of a three-level array structure, as shown below:
```
Comment (spacewander) [inline doc/plugins/fault-injection.md:40]: The `vars` is a list of `expression` which is from the lua-resty-expr.

File: doc/plugins/fault-injection.md [hunk 6]
```diff
@@ -30,11 +30,43 @@ Fault injection plugin, this plugin can be used with other plugins and will be e
...
+
+```json
+array[
```
Comment (spacewander) [inline doc/plugins/fault-injection.md:43]: Better to use a real world example. And this one is too far away from JSON.

File: apisix/plugins/fault-injection.lua [hunk 7]
```diff
@@ -77,11 +151,30 @@ end
 function _M.rewrite(conf, ctx)
     core.log.info("plugin rewrite phase, conf: ", core.json.delay_encode(conf))
 
-    if conf.delay and sample_hit(conf.delay.percentage) then
+    local err
+    local abort_vars = true
+    if conf.abort and conf.abort.vars then
+        abort_vars, err = vars_match(conf.abort.vars, ctx)
+        if err then
+            return 500, err
```
Comment (spacewander) [inline apisix/plugins/fault-injection.lua:159]: Better to log the error.

File: apisix/plugins/fault-injection.lua [hunk 8]
```diff
@@ -77,11 +151,30 @@ end
...
+        end
+    end
+    core.log.info("abort_vars: ", abort_vars)
+
+    local delay_vars = true
+    if conf.delay and conf.delay.vars then
+        delay_vars, err = vars_match(conf.delay.vars, ctx)
+        if err then
+            return 500, err
```
Comment (spacewander) [inline apisix/plugins/fault-injection.lua:168]: Ditto

**Review by spacewander — CHANGES_REQUESTED** [review]

File: apisix/plugins/fault-injection.lua [hunk 9]
```diff
@@ -64,6 +119,25 @@ local function sample_hit(percentage)
...
+            break
+        end
+    end
+
+    return match_result, nil
+end
+
+
 function _M.check_schema(conf)
     local ok, err = core.schema.check(schema, conf)
```
Comment (spacewander) [inline apisix/plugins/fault-injection.lua:142]: Better to check the expression here and add a test for it.

File: apisix/plugins/fault-injection.lua (hunk 1 above)
Comment (Firstsawyou) [reply apisix/plugins/fault-injection.lua:27]: updated.

File: apisix/plugins/fault-injection.lua (hunk 2 above)
Comment (Firstsawyou) [reply apisix/plugins/fault-injection.lua]: updated.

File: doc/plugins/fault-injection.md (hunk 3 above)
Comment (Firstsawyou) [reply doc/plugins/fault-injection.md]: updated

File: apisix/plugins/fault-injection.lua (hunk 9 above)
Comment (Firstsawyou) [reply apisix/plugins/fault-injection.lua:105]: updated.

File: apisix/plugins/fault-injection.lua (hunk 7 above)
Comment (Firstsawyou) [reply apisix/plugins/fault-injection.lua:137]: The log is recorded here: [https://github.com/apache/apisix/pull/3363/files#diff-a2a9ef17c5a2a556f996066b04132d8dd9a848b31eff4a340f5b154283af98deR90 ](https://github.com/apache/apisix/pull/3363/files#diff-a2a9ef17c5a2a556f996066b04132d8dd9a848b31eff4a340f5b154283af98deR90)

File: apisix/plugins/fault-injection.lua (hunk 9 above)
Comment (spacewander) [reply apisix/plugins/fault-injection.lua:105]: No, I mean checking the expression via:
```lua
        local expr, err = expr.new(var)
        if err then
            core.log.error("failed to create vars expression: ", err)
            return nil, err
        end
```

File: apisix/plugins/fault-injection.lua (hunk 1 above)
Comment (spacewander) [reply apisix/plugins/fault-injection.lua:27]: No, you don't remove it.

Comment (spacewander) [reply apisix/plugins/fault-injection.lua:27]: I think just keep:
```
local vars_schema = {
    type = "array",
    maxItems = 20
}
```
is enough. The remain can be validated via `expr.new`.

Comment (Firstsawyou) [reply apisix/plugins/fault-injection.lua:27]: Okay, let me try it.

File: apisix/plugins/fault-injection.lua (hunk 9 above)
Comment (Firstsawyou) [reply apisix/plugins/fault-injection.lua:92]: added.

File: apisix/plugins/fault-injection.lua (hunk 1 above)
Comment (Firstsawyou) [reply apisix/plugins/fault-injection.lua]: updated.

**Review by spacewander — APPROVED** [review]

**Review by membphis — APPROVED** [review]

**Pull request merged**